src/Transports.AspNetCore/GraphQLHttpMiddleware.cs

```diff
@@ -444,17 +449,192 @@ protected virtual async Task<ExecutionResult> ExecuteRequestAsync(HttpContext co
     ValueTask<IDictionary<string, object?>?> IUserContextBuilder.BuildUserContextAsync(HttpContext context, object? payload)
         => BuildUserContextAsync(context, payload);
 
+    private static readonly MediaTypeHeaderValueMs[] _validMediaTypes = new[]
+    {
+        MediaTypeHeaderValueMs.Parse(CONTENTTYPE_GRAPHQLRESPONSEJSON),
+        MediaTypeHeaderValueMs.Parse(CONTENTTYPE_JSON),
+        MediaTypeHeaderValueMs.Parse(CONTENTTYPE_GRAPHQLJSON), // deprecated
+    };
+
     /// <summary>
     /// Selects a response content type string based on the <see cref="HttpContext"/>.
-    /// Defaults to <see cref="CONTENTTYPE_GRAPHQLJSON"/>.  Override this value for compatibility
-    /// with non-conforming GraphQL clients.
+    /// The default implementation attempts to match the content-type requested by the
+    /// client through the 'Accept' HTTP header to the default content type specified
+    /// within <see cref="GraphQLHttpMiddlewareOptions.DefaultResponseContentType"/>.
+    /// If matched, the specified content-type is returned; if not, supported
+    /// content-types are tested ("application/json", "application/graphql+json", and
+    /// "application/graphql-response+json") to see if they match the 'Accept' header.
     /// <br/><br/>
     /// Note that by default, the response will be written as UTF-8 encoded JSON, regardless
-    /// of the content-type value here.  For more complex behavior patterns, override
+    /// of the content-type value here, and this method's default implementation assumes as much.
+    /// For more complex behavior patterns, override
     /// <see cref="WriteJsonResponseAsync{TResult}(HttpContext, HttpStatusCode, TResult)"/>.
     /// </summary>
     protected virtual string SelectResponseContentType(HttpContext context)
-        => _options.ResponseContentType;
+    {
+        // pull the Accept header, which may contain multiple content types
+        var acceptHeaders = context.Request.GetTypedHeaders().Accept;
+
+        if (acceptHeaders != null)
+        {
+            // enumerate through each content type and see if it matches a supported content type
+            // give priority to specific types, then to types with wildcards
+            foreach (var acceptHeader in acceptHeaders.OrderBy(x => x.MatchesAllTypes ? 4 : x.MatchesAllSubTypes ? 3 : x.MatchesAllSubTypesWithoutSuffix ? 2 : 1))
+            {
+                var response = CheckForMatch(acceptHeader);
+                if (response != null)
+                    return response;
+            }
+        }
+
+        // return the default content type if no match is found, or if there is no 'Accept' header
```

Comment:
Hmm, what if client specified "application/pdf" or "application/some_unknown_mime" in accept header?

From https://datatracker.ietf.org/doc/html/rfc7231#section-5.3.2
> A request without any Accept header field implies that the user agent
   will accept any media type in response.  If the header field is
   present in a request and none of the available representations for
   the response have a media type that is listed as acceptable, the
   origin server can either honor the header field by sending a 406 (Not
   Acceptable) response **or disregard the header field by treating the
   response as if it is not subject to content negotiation.**


Do you want to go with second option?

Reply:
That’s the case now (merged in this PR). Please read my response to your comments on this earlier today. TLDR: maybe update in v8 to reject with 406